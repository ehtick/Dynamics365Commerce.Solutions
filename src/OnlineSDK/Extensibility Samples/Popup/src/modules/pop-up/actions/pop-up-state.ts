/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */

/**
 * State information related to handle Pop up.
 */
export interface IPopUpState {
    /**
     * Set this to true to communicate that Pop up dialog should open and vice versa.
     */
    isOpen?: boolean;
}
