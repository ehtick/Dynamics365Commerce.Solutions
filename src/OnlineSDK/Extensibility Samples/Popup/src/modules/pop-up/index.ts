/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */

export * from './pop-up';
export * from './pop-up.props.autogenerated';
export * from './pop-up.view';
export * from './actions';
