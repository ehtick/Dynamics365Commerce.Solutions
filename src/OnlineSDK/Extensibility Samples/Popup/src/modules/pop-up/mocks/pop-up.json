{
  "id": "PopUpModule",
  "config": {
    "heading": "Subscribe"
  },
  "typeName": "pop-up",
  "modules": {
    "content": [
      {
        "typeName": "text-block",
        "id": "text-block",
        "friendlyName": "Text block",
        "config": {
          "paragraph": "<p><u><a aria-label=\"My account\" href=\"/myaccount\">My account</a></u> /<a href=\"/orderhistory\">&nbsp;<u>Order history</u></a>/&nbsp;Order details</p>\n\n<p>&nbsp;</p>\n",
          "className": "breadcrumb"
        }
      },
      {
        "typeName": "subscribe",
        "id": "IDNLAkOv_jrna7iv-subscribe",
        "friendlyName": "Subscribe 1",
        "config": {
            "heading": {
                "text": "Sign up for AW newsletter",
                "tag": "h2"
            },
            "paragraph": "<p>Latest trends, styles &amp; promotions. Are you on the list? Subscribe and get the latest hot deals!</p>\n"
        }
    }
    ]
  }
}