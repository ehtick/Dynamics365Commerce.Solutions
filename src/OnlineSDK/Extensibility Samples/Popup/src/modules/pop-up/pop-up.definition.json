{
    "$type": "containerModule",
    "friendlyName": "Pop Up",
    "name": "pop-up",
    "description": "The pop-up module will create a modal dialog and attach it to the DOM. Other modules could be exposed inside this module to make them appear inside the pop-up",
    "categories": ["pop-up"],
    "slots": {
        "content":{
            "friendlyName": "PopUp item slot",
            "description": "PopUp item slot",
            "allowedTypes": "*",
            "min": 1,
            "max": 10
        }
    },
    "config": {
        "heading": {
            "friendlyName": "Heading",
            "description": "Provide a heading",
            "type": "string",
            "default": ""
        },
        "className": {
            "friendlyName": "Custom CSS class name",
            "description": "Provide a custom css class name defined in the theme pack to target the CSS for this module",
            "type": "string",
            "default": ""
        },
        "isBackDropStatic": {
            "friendlyName": "Option to choose a static backdrop. Static backdrop doesn't close modal on click",
            "description": "Provide true if no click action is desired in the backdrop",
            "type": "boolean",
            "default": false
        },
        "isCookieEnabled": {
            "friendlyName": "Option to associate pop-up view based on the cookie",
            "description": "Provide true if pop-up display should depend on cookie",
            "type": "boolean",
            "default": false
        },
        "shouldPopUpOnLoad": {
            "friendlyName": "Option to choose if the pop-up should pop up on initial load",
            "description": "Provide true if pop-up should pop up on initial load",
            "type": "boolean",
            "default": false
        }
    },
    "resources": {
        "popUpAriaLabel": {
            "value": "Pop Up",
            "comment": "Aria label for Pop Up"
        }
    },
    "dataActions": {
        "popUpState": {
            "path": "./actions/pop-up-state-data-action"
        }
    }
}