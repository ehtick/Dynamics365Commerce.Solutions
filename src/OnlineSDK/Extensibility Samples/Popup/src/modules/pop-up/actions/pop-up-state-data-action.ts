/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */

import { createObservableDataAction, IAction, IActionInput } from '@msdyn365-commerce/core';
import { GenericInput, getGenericAction } from '@msdyn365-commerce-modules/retail-actions';

import { IPopUpState } from './pop-up-state';

/**
 * This method builds the input for the createPopUpStateInput call.
 * @param result - Popup open state.
 * @returns - New GenericInput.
 */
export function setPopUpState(result: IPopUpState): GenericInput<IPopUpState> {
    return new GenericInput<IPopUpState>('popUpState', result, 'IPopUpState');
}

/**
 * This method returns the input for the createPopUpOpenStateInput call.
 * @returns - New Popup state input.
 */
const createPopUpStateInput = (): IActionInput => {
    return setPopUpState({});
};

/**
 * Generic data action.
 */
export const getGenericActionDataAction = createObservableDataAction({
    action: getGenericAction as IAction<IPopUpState>,
    input: createPopUpStateInput
});

export default getGenericActionDataAction;
