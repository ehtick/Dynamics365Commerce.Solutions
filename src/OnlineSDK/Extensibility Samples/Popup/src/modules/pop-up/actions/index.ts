/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */

import { IPopUpState } from './pop-up-state';
import { setPopUpState } from './pop-up-state-data-action';

/**
 * Export Pop up data action.
 */
export { IPopUpState, setPopUpState };
