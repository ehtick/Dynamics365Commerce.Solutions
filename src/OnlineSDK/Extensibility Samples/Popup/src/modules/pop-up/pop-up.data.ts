/*--------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * See License.txt in the project root for license information.
 *--------------------------------------------------------------*/

import { AsyncResult } from '@msdyn365-commerce/core';

import { IPopUpState } from './actions';

/**
 * Pop up data.
 */
export interface IPopUpData {
    popUpState: AsyncResult<IPopUpState>;
}
