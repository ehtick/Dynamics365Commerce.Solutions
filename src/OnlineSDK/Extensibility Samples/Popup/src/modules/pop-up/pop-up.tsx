/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */

import * as Msdyn365 from '@msdyn365-commerce/core';
import { IModuleProps, INodeProps, Modal, ModalBody, ModalHeader } from '@msdyn365-commerce-modules/utilities';
import classnames from 'classnames';
import { observer } from 'mobx-react';
import * as React from 'react';

import { setPopUpState } from './actions';
import { IPopUpData } from './pop-up.data';
import { IPopUpProps } from './pop-up.props.autogenerated';

/**
 * Pop up view props.
 */
export interface IPopUpViewProps extends IPopUpProps<IPopUpData> {
    moduleProps: IModuleProps;
    headerContainerProps: INodeProps;
    headerProps?: React.ReactNode;
    bodyContainerProps: INodeProps;
    bodyContent: React.ReactNode;
    className: string;
    onDismiss(): void;
}

/**
 *
 * PopUp component.
 * @extends {React.PureComponent<IPopUpProps<IPopUpData>>}
 */
@observer
class PopUp extends React.PureComponent<IPopUpProps<IPopUpData>> {
    public componentDidMount(): void {
        const {
            config: { isCookieEnabled, shouldPopUpOnLoad },
            context: {
                actionContext,
                app: {
                    config: { popUpCookieName }
                }
            }
        } = this.props;

        if (shouldPopUpOnLoad) {
            const popUpState = { isOpen: true };
            const cookieName = (popUpCookieName as string) || '_msdyn365__popUp_';
            if (!isCookieEnabled || !this._isActionTaken(cookieName)) {
                actionContext.update(setPopUpState(popUpState), popUpState);
            }
        }
    }

    public render(): JSX.Element | null {
        const {
            config: { className, heading, isBackDropStatic },
            data,
            slots: { content },
            resources
        } = this.props;

        const isDialogOpen = data.popUpState.result?.isOpen;

        const viewProps = {
            ...this.props,
            onDismiss: this._onDismiss,
            moduleProps: {
                tag: Modal,
                moduleProps: this.props,
                className: classnames('ms-pop-up', className),
                autoFocus: true,
                backdrop: isBackDropStatic ? 'static' : true,
                fade: true,
                isOpen: isDialogOpen,
                'aria-label': heading ?? resources.popUpAriaLabel,
                onClosed: this._onDismiss,
                horizontalPosition: 'center',
                verticalPosition: 'center',
                toggle: this._onDismiss
            },
            headerContainerProps: {
                tag: ModalHeader,
                className: 'ms-pop-up__modal-header',
                toggle: this._onDismiss
            },
            headerProps: heading && (
                <Msdyn365.Text
                    className='ms-pop-up__modal-title'
                    {...heading}
                    tag='h2'
                    text={heading}
                    editProps={{ onEdit: this.handleHeadingChange, requestContext: this.props.context.request }}
                />
            ),
            bodyContainerProps: {
                tag: ModalBody,
                className: 'ms-pop-up__modal-body'
            },
            bodyContent: content.map((item: React.ReactNode, index: number) => this._getPopUpContent(item, index))
        };

        return this.props.renderView(viewProps) as React.ReactElement;
    }

    /**
     * Handle the heading change event.
     * @param event - Content Editable Event.
     */
    public handleHeadingChange = (event: Msdyn365.ContentEditableEvent): void => {
        this.props.config.heading = event.target.value;
    };

    /**
     * _isActionTaken method - To check cookie value.
     * @param cookieName - Cookie name.
     * @returns - Boolean value for existence of cookie or it's value itself.
     */
    private readonly _isActionTaken = (cookieName: string): boolean => {
        const cookie = this.props.context.request.cookies;
        const popUpCookie = cookie.get(cookieName) as Msdyn365.ICookieValue<{}> | undefined;

        if (popUpCookie === undefined || popUpCookie.value === 'false') {
            return false;
        }

        return !!popUpCookie.value;
    };

    /**
     * _onDismiss method - To dismiss pop up.
     */
    private readonly _onDismiss = (): void => {
        const popUpState = { isOpen: false };
        this.props.context.actionContext.update(setPopUpState(popUpState), popUpState);
    };

    /**
     * _getPopUpContent method - To get pop up content.
     * @param item - Pop up item.
     * @param index - Index.
     * @returns - Pop up content as react element.
     */
    private readonly _getPopUpContent = (item: React.ReactNode, index: number): React.ReactNode => {
        return <React.Fragment key={index}>{React.cloneElement(item as React.ReactElement)}</React.Fragment>;
    };
}

export default PopUp;
