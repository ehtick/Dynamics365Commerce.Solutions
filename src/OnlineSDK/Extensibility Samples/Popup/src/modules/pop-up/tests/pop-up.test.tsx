/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */

import * as Msdyn365 from '@msdyn365-commerce/core';
import { buildMockModuleProps } from '@msdyn365-commerce/core-internal';
import { wrapInResolvedAsyncResult } from '@msdyn365-commerce-modules/retail-actions';
import { mount } from 'enzyme';
import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { IPopUpState } from '../actions';
import PopUp from '../pop-up';
import { IPopUpData } from '../pop-up.data';
import { IPopUpConfig, IPopUpProps, IPopUpResources } from '../pop-up.props.autogenerated';

const mockData: IPopUpData = {
    popUpState: wrapInResolvedAsyncResult({
        isOpen: true
    } as IPopUpState)
};

const mockConfig: IPopUpConfig = {
    heading: 'Subscribe'
};

const mockActions = {};

const mockSlots = {
    content: [<div key='1' />]
};

const mockResources: IPopUpResources = {
    popUpAriaLabel: 'Test'
};

describe('PopUp', () => {
    let moduleProps: IPopUpProps<IPopUpData>;

    it('renders correctly', () => {
        moduleProps = buildMockModuleProps(mockData, mockActions, mockConfig) as IPopUpProps<IPopUpData>;
        moduleProps.slots = mockSlots;
        moduleProps.resources = mockResources;
        const component: renderer.ReactTestRenderer = renderer.create(<PopUp {...moduleProps} />);
        const tree: renderer.ReactTestRendererJSON | renderer.ReactTestRendererJSON[] | null = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly with out heading', () => {
        moduleProps = buildMockModuleProps(mockData, mockActions, mockConfig) as IPopUpProps<IPopUpData>;
        moduleProps.slots = mockSlots;
        moduleProps.resources = mockResources;
        moduleProps.config.heading = undefined;
        const component: renderer.ReactTestRenderer = renderer.create(<PopUp {...moduleProps} />);
        const tree: renderer.ReactTestRendererJSON | renderer.ReactTestRendererJSON[] | null = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('handles heading change event', () => {
        moduleProps = buildMockModuleProps(mockData, mockActions, mockConfig) as IPopUpProps<IPopUpData>;
        moduleProps.slots = mockSlots;
        moduleProps.resources = mockResources;
        moduleProps.config.heading = undefined;
        const component = mount(<PopUp {...moduleProps} />);
        const popUpInstance = component.instance() as PopUp;
        const event = { currentTarget: { value: 'foo1' }, target: { value: 'foo' } } as Msdyn365.ContentEditableEvent;

        popUpInstance.handleHeadingChange(event);

        expect.assertions(0);
    });

    it('checks for cookie', () => {
        moduleProps = buildMockModuleProps(mockData, mockActions, mockConfig) as IPopUpProps<IPopUpData>;
        moduleProps.slots = mockSlots;
        moduleProps.resources = mockResources;
        moduleProps.config.heading = undefined;
        const component = mount(<PopUp {...moduleProps} />);
        const popUp = component.instance() as PopUp;

        // @ts-expect-error ignore private method
        const isActionTaken = popUp._isActionTaken('Random_Cookie_Name_XYZ');
        expect(isActionTaken).toBe(false);
    });

    it('returns popUp Content', () => {
        moduleProps = buildMockModuleProps(mockData, mockActions, mockConfig) as IPopUpProps<IPopUpData>;
        moduleProps.slots = mockSlots;
        moduleProps.resources = mockResources;
        moduleProps.config.heading = undefined;
        const component = mount(<PopUp {...moduleProps} />);
        const popUp = component.instance() as PopUp;

        // @ts-expect-error ignore private method
        const item = popUp._getPopUpContent(moduleProps.slots.content[0], 0);
        expect(item).toBeTruthy();
    });
});
