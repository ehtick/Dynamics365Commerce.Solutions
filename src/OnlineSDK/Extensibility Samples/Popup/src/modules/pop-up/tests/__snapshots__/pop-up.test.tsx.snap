// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopUp renders correctly 1`] = `null`;

exports[`PopUp renders correctly with out heading 1`] = `null`;
